File: leetcode/0707.design-linked-list/go/001.go
type Node struct {
        Val  int
        Next *Node
}

type MyLinkedList struct {
        head   *Node
        length int
}

func Constructor() MyLinkedList {
        return MyLinkedList{}
}

func (this *MyLinkedList) Get(index int) int {

        if index < 0 || index > this.length-1 {
                return -1
        }

        cur := this.head
        for i := 0; i < index && cur != nil; i++ {
                cur = cur.Next
        }

        if cur == nil {
                return -1
        }

        return cur.Val
}

func (this *MyLinkedList) AddAtHead(val int) {
        var newNode Node
        newNode.Val = val
        newNode.Next = this.head
        this.head = &newNode

        this.length++
}

func (this *MyLinkedList) AddAtTail(val int) {
        if this.head == nil {
                this.AddAtHead(val)
                return
        }

        cur := this.head
        for i := 0; cur.Next != nil; i++ {
                cur = cur.Next
        }

        var newNode Node
        newNode.Val = val
        cur.Next = &newNode

        this.length++
}

func (this *MyLinkedList) AddAtIndex(index int, val int) {
        if index == 0 {
                this.AddAtHead(val)
                return
        } else if index == this.length {
                this.AddAtTail(val)
                return
        } else if index < 0 || index > this.length {
                return
        }

        cur := this.head
        for i := 0; i < index-1; i++ {
                cur = cur.Next
        }

        var newNode Node
        newNode.Val = val
        newNode.Next = cur.Next
        cur.Next = &newNode

        this.length++

}

func (this *MyLinkedList) DeleteAtIndex(index int) {
        if index < 0 || index >= this.length {
                return
        } else if index == 0 {
                this.head = this.head.Next
                this.length--
                return
        }

        cur := this.head
        for i := 0; i < index-1; i++ {
                cur = cur.Next
        }

        cur.Next = cur.Next.Next
        this.length--

}


/**
 * Your MyLinkedList object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Get(index);
 * obj.AddAtHead(val);
 * obj.AddAtTail(val);
 * obj.AddAtIndex(index,val);
 * obj.DeleteAtIndex(index);
 */

File: leetcode/0707.design-linked-list/javascript/001.js
class ListNode {
  constructor(val, next) {
    this.val = val == undefined ? 0 : val;
    this.next = next == undefined ? null : next;
  }
}

var MyLinkedList = function() {
    this.head = null;
};

MyLinkedList.prototype.get = function(index) {
    if (index < 0 || index >= this.getLength()) return -1;
    let cur = this.head;
    for (let i = 0; i < index; i++) {
        cur = cur.next;
    }
    return cur.val;
};

MyLinkedList.prototype.addAtHead = function(val) {
    let node = new ListNode(val);
    node.next = this.head;
    this.head = node;
};

MyLinkedList.prototype.addAtTail = function(val) {
    if (this.head == null) {
        this.addAtHead(val);
        return;
    }
    let node = new ListNode(val);
    let cur = this.head;
    while (cur.next !== null) {
        cur = cur.next;
    }
    cur.next = node;
};

MyLinkedList.prototype.addAtIndex = function(index, val) {
    if (index === 0) {
        this.addAtHead(val);
        return;
    }
    if (index === this.getLength()) {
        this.addAtTail(val);
        return;
    }
    if (index > this.getLength()) return;
        
    let cur = this.head;
    let node = new ListNode(val);
    for (let i = 0; i < index-1; i++) {
        cur = cur.next;
    }
    let next = cur.next;
    cur.next = node;
    node.next = next;
};

MyLinkedList.prototype.deleteAtIndex = function(index) {
    if (index < 0 || index >= this.getLength()) return;
    if (index == 0) {
        this.head = this.head.next;
        return;
    }

    let cur = this.head;
    for (let i = 0; i < index-1; i++) {
        cur = cur.next;
    }
    cur.next = cur.next.next;
};

MyLinkedList.prototype.getLength = function() {
    let len = 0, cur = this.head;
    while (cur !== null) {
        cur = cur.next;
        len++;
    }
    return len;
};
